### PR TITLE
Supporting the __dunder__ addition and exponentiation methods for operators!

### PR DESCRIPTION
**Pull request by AlbertMitjans**
**Context:**

Added support so that users can combine operators more naturally by implementing the dunder methods `__add__` and `__pow__` at the level of `Operator`:

```python
>>> summed_op = qml.RX(phi=1.23, wires=0) + qml.RZ(phi=3.14, wires=0)
>>> summed_op
RX(1.23, wires=[0]) + RZ(3.14, wires=[0])
>>> exp_op = qml.RZ(1.0, wires=0) ** 2 
>>> exp_op
RZ**2(1.0, wires=[0])
```

**Description of the Change:**
- `changelog-dev.md`:
  - Add info of the new feature.
  - Add my name in the contributors list.
  - Fix trailing spaces and other markdown style issues.
- `__init__.py`:
  - Import `Sum` and `Pow` explicitly given that (for some unknown reason to me) they are not imported when using `from pennylane.ops import *`.
- `operation.py`:
  - Add `__add__` and `__pow__` dunder methods to `Operation` class.
  - Call `super().__add__` method before raising error in `Observable` class.
  - Use `copy.copy` instead of `self.__copy__()` to avoid pylint error.
  - Sort imports.
- `sum.py`:
  - Call `summand.summands` when defining `self.summands` if `summand` is a `Sum` instance.
  - Define type of `summands` in `__init__` call.
  - Sort imports.
- `hamiltonian.py`:
  - Cast `coeffs` to list to be consistent with the outputs of the `terms()` method of the `Sum` class.
- `vqe.py`:
  - Cast `coeffs` to tuple because `group_observables` casts the output to `list` if input is `list`: https://github.com/PennyLaneAI/pennylane/blob/c8f78b632bfaced6806f0d5e61b09a84fa365a15/pennylane/grouping/group_observables.py#L263
  - Why is this done?
  - Remove useless list comprehension.
  - Sort imports.
- `test_pow_op.py` + `test_sum.py`:
  - Test all tests with using both class instantiation and the `Operation` dunder method.
  - Theoretically it makes more sense to add the tests in the `test_operation.py` file (given that the new features are located in this class), but I finally decided to add them in these files to make sure that all the corner cases are tested.
  - Some tests are targeted to class instantiation explicitly. In these cases we use only the class instantiation method.
  - Using the `+` operator can instantiate either a `Sum` or a `Hamiltonian` class (depending on the arguments). Given that some tests use specific methods of the `Sum` class, we need to check that the return value is an instance of this class.
- `test_hamiltonian.py`:
  - Changed the `coeff` instance check to `list`.
- `test_operation.py`:
  - Add test checking that an error is raised when raising an operator to a power that is not a number.
  - Sort imports.

**Benefits:**  
Users can combine operators more naturally.

**Possible Drawbacks:** 
Users cannot define the `do_queue` and `id` arguments of the `Sum` and `Pow` class when using the dunder methods.

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/issues/2803

### CODE REVIEW

File: pennylane/ops/op_math/sum.py [hunk 1]
```diff
@@ -128,9 +131,15 @@ def __init__(self, *summands, do_queue=True, id=None):  # pylint: disable=super-
...
-        self._wires = qml.wires.Wires.all_wires([s.wires for s in summands])
+        self.summands: List[Operator] = []
+        for summand in summands:
+            if isinstance(summand, Sum):
+                self.summands.extend(summand.summands)
+                continue
+            self.summands.append(summand)
+
+        self.data = [s.parameters for s in self.summands]
+        self._wires = qml.wires.Wires.all_wires([s.wires for s in self.summands])
```
Comment (Jaybsoni) [inline pennylane/ops/op_math/sum.py:142]: I think for now we can leave the simplification logic out. This should be addressed by a separate PR, and we still need to figure out if we want to simplify during each construction automatically, or if its better to provide a top level function to simplify in general 👍🏼 

```suggestion
        self.summands = summands
        self.data = [s.parameters for s in self.summands]
        self._wires = qml.wires.Wires.all_wires([s.wires for s in self.summands])
```

File: pennylane/vqe/vqe.py [hunk 2]
```diff
@@ -200,10 +204,7 @@ def circuit(*qnode_args, obs, **qnode_kwargs):
             def cost_fn(*qnode_args, **qnode_kwargs):
                 """Combine results from grouped QNode executions with grouped coefficients"""
                 if device.shot_vector:
-                    shots_batch = 0
-
-                    for i in device.shot_vector:
-                        shots_batch += i[1]
+                    shots_batch = sum(i[1] for i in device.shot_vector)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/vqe/vqe.py:207]: Haha! I was going to say, some of the changes suggested felt very stylistic / lint-y. What IDE are you using?

[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/vqe/vqe.py:207]: VS Code!

**Comment by Jaybsoni** [comment]
[sc-20265]

**Comment by AlbertMitjans** [comment]
[sc-22759]

**Review by Jaybsoni — APPROVED** [review]
Looks good to merge 💯!  Great work on your first contribution !
